src/node/state.go

```diff
@@ -34,37 +52,54 @@ func (s NodeState) String() string {
 	}
 }
 
-type nodeState struct {
-	state    NodeState
-	wg       sync.WaitGroup
-	locker   sync.Mutex
+func (s *nodeState2) mtx() {
+	for {
+		select {
+		case s.state = <-s.setStateChan:
+		case s.getStateChan <- s.state:
+		}
+	}
 }
 
-func (b *nodeState) getState() NodeState {
-	stateAddr := (*uint32)(&b.state)
-	return NodeState(atomic.LoadUint32(stateAddr))
-}
+func (s *nodeState2) goFunc(fu func()) {
+	go func() {
+		s.lock.Lock()
+		s.wip++
+		s.lock.Unlock()
+
+		fu()
+
+		s.lock.Lock()
+		s.wip--
+		s.lock.Unlock()
 
-func (b *nodeState) setState(s NodeState) {
-	stateAddr := (*uint32)(&b.state)
-	atomic.StoreUint32(stateAddr, uint32(s))
+		s.cond.L.Lock()
+		defer s.cond.L.Unlock()
+		s.cond.Broadcast()
+	}()
 }
 
-// Start a goroutine and add it to waitgroup
-func (b *nodeState) goFunc(f func()) {
-	b.locker.Lock()
-	defer b.locker.Unlock()
-	if b.getState() != Shutdown {
-		b.wg.Add(1)
-		go func() {
-			defer b.wg.Done()
-			f()
-		}()
+func (s *nodeState2) waitRoutines() {
+	s.cond.L.Lock()
+	defer s.cond.L.Unlock()
+
+	for {
+		s.lock.RLock()
+		wip := s.wip
+		s.lock.RUnlock()
+
+		if wip != 0 {
+			s.cond.Wait()
+			continue
+		}
+		break
```

Comment:
At this point a new GoFunc() can be started on heartbeat or if any other node decides to make a sync with current node... and this ensures complete failure of waitRoutines() in this implementation.

Reply:
To fix this case, we need a big refactoring of `src / node` pkg.  I do not like the current architecture. In this PR, I focused on fixing the data race for Node.state () without affecting the architecture.
Please see the description of PR.

Reply:
This case has nothing to do with node package. It's can be solved within nodestate.
What is the point in rushing to merge in knowingly flawed code?